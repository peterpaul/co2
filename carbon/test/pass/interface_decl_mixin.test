/*
 * carbon Compiler and programming language for libco2
 *
 * Copyright (C) 2011,2012 Peterpaul Taekele Klein Haneveld
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
["stdio.h"] int printf(char * format, ...);

interface Foo {
	int foo (int a);
	int bar (Foo f) {
		return f.foo (10) + foo (5);
	}
}

class Bar, Foo {
	int foo (int a) {
		return a + 2;
	}
}

class FooBar, Foo {
	int foo (int a) {
		return a - 2;
	}

	int bar (Foo f) {
		return f.foo (10) - foo (5);
	}
}

int main () {
	Foo f = new Bar ();
	Foo g = new FooBar ();
	printf ("f.bar(f) = %d\n", f.bar (f)); // (10 + 2) + (5 + 2) = 19
	printf ("f.bar(g) = %d\n", f.bar (g)); // (10 - 2) + (5 + 2) = 15
	printf ("g.bar(f) = %d\n", g.bar (f)); // (10 + 2) - (5 - 2) = 9
	printf ("g.bar(g) = %d\n", g.bar (g)); // (10 - 2) - (5 - 2) = 5
	return 0;
}
