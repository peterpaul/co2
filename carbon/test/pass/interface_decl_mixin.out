f.bar(f) = 19
f.bar(g) = 15
g.bar(f) = 9
g.bar(g) = 5
